File: 1/2020.09.25/264.Ugly Number II/java/solution.java
class Solution {
    public int nthUglyNumber(int n) {
        List<Integer> ugly = new ArrayList<>();
        ugly.add(1);
        int i2 = 0, i3 = 0, i5 = 0;
        while (n > 1) {
            int u2 = 2 * ugly.get(i2);
            int u3 = 3 * ugly.get(i3);
            int u5 = 5 * ugly.get(i5);
            int umin = Math.min(u2, Math.min(u3, u5));
            if (umin == u2) {
                i2++;
            }
            if (umin == u3) {
                i3++;
            }
            if (umin == u5) {
                i5++;
            }
            ugly.add(umin);
            n--;
        }
        return ugly.get(ugly.size() - 1);
    }
}

File: 1/2020.09.25/264.Ugly Number II/python/solution.py
from heapq import heappop, heappush
class Solution:
    def nthUglyNumber(self, n: int) -> int:
        ugly = [1]
        i2 = i3 = i5 = 0
        while n > 1:
            u2, u3, u5 = 2*ugly[i2], 3*ugly[i3], 5*ugly[i5]
            umin = min(u2, u3, u5)
            if umin == u2:
                i2 += 1
            if umin == u3:
                i3 += 1
            if umin == u5:
                i5 += 1
            ugly.append(umin)
            n -= 1
        return ugly[-1]

    def nthUglyNumber(self, n: int) -> int:
        heap = []
        count = 0
        heappush(heap, 1)
        while True:
            res = heappop(heap)
            count += 1
            if count == n:
                return res
            while heap != [] and heap[0] == res:
                heappop(heap)
            heappush(heap, res*2)
            heappush(heap, res*3)
            heappush(heap, res*5)
